src/core/net/ip6.cpp

```diff
@@ -603,11 +614,303 @@ otError Ip6::HandleOptions(Message &aMessage, Header &aHeader, bool &aForward)
+        assertValue = message->Write(0, sizeof(header), &header);
+        assert(assertValue == sizeof(header));
+
+        otLogDebgIp6("Reassembly complete.");
+
+        mReassemblyList.Dequeue(*message);
+
+        error = HandleDatagram(*message, aNetif, aMessageInfo.mLinkInfo, aFromNcpHost);
+    }
+
+exit:
+    if (error != OT_ERROR_DROP && error != OT_ERROR_NONE && isFragmented)
+    {
+        if (message != NULL)
+        {
+            mReassemblyList.Dequeue(*message);
+            message->Free();
+        }
+        otLogWarnIp6("Reassembly failed: %s", otThreadErrorToString(error));
+    }
+    if (isFragmented)
+    {
+        // drop all fragments, the payload is stored in the fragment buffer
+        error = OT_ERROR_DROP;
+    }
+
+    return error;
+}
+
+void Ip6::CleanupFragmentationBuffer(void)
+{
+    Message *message;
+    Message *next;
+
+    for (message = mReassemblyList.GetHead(); message; message = next)
+    {
+        next = message->GetNext();
+        mReassemblyList.Dequeue(*message);
+
+        message->Free();
+    }
+}
+
+void Ip6::HandleTimer(Timer &aTimer)
+{
+    aTimer.GetOwner<Ip6>().HandleUpdateTimer();
+}
+
+void Ip6::HandleUpdateTimer(void)
+{
+    if (UpdateReassemblyList())
+    {
+        mTimer.Start(kStateUpdatePeriod);
+    }
+}
+
+bool Ip6::UpdateReassemblyList(void)
+{
+    Message *next;
+
+    for (Message *message = mReassemblyList.GetHead(); message; message = next)
+    {
+        next = message->GetNext();
+
+        if (message->GetTimeout() > 0)
+        {
+            message->DecrementTimeout();
+        }
+        else
+        {
+            otLogNoteIp6("Reassembly timeout.");
+            SendIcmpError(*message, IcmpHeader::kTypeTimeExceeded, IcmpHeader::kCodeFragmReasTimeEx);
+
+            mReassemblyList.Dequeue(*message);
+            message->Free();
+        }
+    }
+    return mReassemblyList.GetHead() != NULL;
+}
+
+otError Ip6::SendIcmpError(Message &aMessage, IcmpHeader::Type aIcmpType, IcmpHeader::Code aIcmpCode)
+{
+    otError     error = OT_ERROR_NONE;
+    Header      header;
+    MessageInfo messageInfo;
+
+    VerifyOrExit(aMessage.Read(0, sizeof(header), &header) == sizeof(header), error = OT_ERROR_PARSE);
+
+    messageInfo.SetPeerAddr(header.GetSource());
+    messageInfo.SetSockAddr(header.GetDestination());
+    messageInfo.SetHopLimit(header.GetHopLimit());
+    messageInfo.SetLinkInfo(NULL);
+
+    SuccessOrExit(error = mIcmp.SendError(aIcmpType, aIcmpCode, messageInfo, header));
+
+exit:
+    return error;
+}
+
+#else
+otError Ip6::HandleFragmentation(Message &aMessage, IpProto aIpProto)
+{
+    EnqueueDatagram(aMessage);
+
+    OT_UNUSED_VARIABLE(aIpProto);
+
+    return OT_ERROR_NONE;
+}
+
+otError Ip6::HandleFragment(Message &aMessage, Netif *aNetif, MessageInfo &aMessageInfo, bool aFromNcpHost)
 {
     otError        error = OT_ERROR_NONE;
     FragmentHeader fragmentHeader;
 
+    OT_UNUSED_VARIABLE(aNetif);
```

Comment:
nit: prefer to move this at top of method.

Reply:
done